### ISSUE TITLE
Question: Pork chop plot generator

### ISSUE THREAD

**Issue by ChristopherRabotin** [poliastro/poliastro#156]
I'm looking to validate my Lambert solver and my PCP (cf. https://github.com/ChristopherRabotin/smd ), so I was wondering whether this tool already supported PCPs or not.

Thanks

**Comment by Juanlu001** [comment]
Hi @ChristopherRabotin, it's been in the back of my mind for a while but never really implemented it. I'm leaving this as a future enhancement.

Pinging @eramirem here for the record.

**Comment by eramirem** [comment]
@ChristopherRabotin @Juanlu001 Noted. Will work on this soon.

**Comment by ZachDischner** [comment]
Adding my interest for this too. I ended up making my own porkchop plotter built on top of Poliastro a while back, but built in 'chops would be great!

Note, this isn't working right now, have to update it to work with newer Poliastro releases. 
https://github.com/ZachDischner/spaceypy

**Comment by Juanlu001** [comment]
Hi @ZachDischner, that's amazing! I didn't know people were already creating things _on top of_ poliastro, so I went ahead and created a section in the wiki for that:

https://github.com/poliastro/poliastro/wiki/Projects-using-poliastro

As there's so much interest for this feature, I'm advancing it for the next milestone so we can all better track its progress :)

**Comment by ZachDischner** [comment]
@Juanlu001 Great! Thanks for the mention! Poliastro is sweet, I keep hoping to find more chances to use it. Keep up the good work!

**Comment by Juanlu001** [comment]
I'm moving this off the 0.7 milestone. In case any of you still want to implement this though, feel free to send a pull request at any time :)

**Comment by Juanlu001** [comment]
Just discovered this beauty:

http://sdg.aero.upm.es/index.php/online-apps/porkchop-plot

![screenshot_2018-08-15 easy porkchop](https://user-images.githubusercontent.com/316517/44169723-a1ede200-a0d5-11e8-8ad0-60e68b792a47.png)

(via https://space.stackexchange.com/a/30120/10716)

However, it's distributed under GPLv3 or later, so **don't use this code as inspiration** to implement this algorithm in poliastro ([more info](https://www.reddit.com/r/programming/comments/94446/ask_proggit_porting_a_gpl_project_to_a_new/)).

**Comment by ZachDischner** [comment]
Damn, nice! This is quick!


On March 24, 2017 at 12:06:18 PM, Juan Luis Cano Rodríguez (
notifications@github.com) wrote:

Hi @ZachDischner <https://github.com/ZachDischner>, that's amazing! I
didn't know people were already creating things *on top of* poliastro, so I
went ahead and created a section in the wiki for that:

https://github.com/poliastro/poliastro/wiki/Projects-using-poliastro

As there's so much interest for this feature, I'm advancing it for the next
milestone so we can all better track its progress :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/poliastro/poliastro/issues/156#issuecomment-289101697>,
or mute the thread

**Comment by ZachDischner** [comment]
i've been wanting to experiment with putting Poliastro up on AWS and
exposing orbit calculations through AWS Lambda and API gateway, would be a
fun tech project. Didn't get very far once I figured out I couldn't just
bundle up the python env for lambda, would have to have a long running
service spun up.

Anyways, thanks for sharing!


On March 24, 2017 at 12:06:18 PM, Juan Luis Cano Rodríguez (
notifications@github.com) wrote:

Hi @ZachDischner <https://github.com/ZachDischner>, that's amazing! I
didn't know people were already creating things *on top of* poliastro, so I
went ahead and created a section in the wiki for that:

https://github.com/poliastro/poliastro/wiki/Projects-using-poliastro

As there's so much interest for this feature, I'm advancing it for the next
milestone so we can all better track its progress :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/poliastro/poliastro/issues/156#issuecomment-289101697>,
or mute the thread

**Comment by ZachDischner** [comment]
Lambda wants you to upload a zip of the whole environment to use during
runtime. Poliastro dependencies make the zip too big for their size limits
(and it is a pain in the butt to build libs for Amazon Linux). I figured
the best path would be to get poliastro and a wrapping flask REST server to
run in a docker container, then run the container as a service using
ElasticContainerService. Then expose that via lambda to make porkchops,
return delta v's, save plots in S3 or online, etc.

Just became more involved and I haven't devoted much time to it. Definitely
something I'm interested in pursuing though!


On August 15, 2018 at 2:21:50 PM, Juan Luis Cano Rodríguez (
notifications@github.com) wrote:

@ZachDischner <https://github.com/ZachDischner> what specific problems did
you have in AWS Lambda to create a proper Python environment with
poliastro? Perhaps we can work them out!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/poliastro/poliastro/issues/156#issuecomment-413323251>,
or mute the thread

**Comment by Juanlu001** [comment]
Let's continue that discussion in the other issue to keep this one focused on porkchop plots :)

**Comment by ZachDischner** [comment]
Ooh that is awesome! Killer interactivity and visualizations.

It'd be a nice separation to have a lightweight static web interface with
poliastro and a js plotter so that all Poliastro runs is the backend
calculations. Communicates with a service call so you could use a local
service, or one in a cloud or on a server somewhere. Even serve the site
statically on AWS...


On September 4, 2018 at 10:23:28 AM, Juan Luis Cano Rodríguez (
notifications@github.com) wrote:

And this other beauty!

https://ssd.jpl.nasa.gov/?mdesign_server&des=433

[image: screenshot_2018-09-04 jpl mission design tool]
<https://user-images.githubusercontent.com/316517/45044161-9c017600-b06f-11e8-87e0-3329b9ea98fd.png>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/poliastro/poliastro/issues/156#issuecomment-418431452>,
or mute the thread

**Comment by Juanlu001** [comment]
> It'd be a nice separation to have a lightweight static web interface with poliastro and a js plotter so that all Poliastro runs is the backend calculations.

Created another issue for that: https://github.com/poliastro/poliastro/issues/444

**Comment by piscisvolans** [comment]
Hi all, since the discussion about porkchop plots seems to be ongoing, I wanted to mention that I have also written a porkchop-plotter based on poliastro. It can generate porkchops between two solar system bodies (e.g., Earth and Mars) or between two arbitrary poliastro Orbit objects (implemented specifically to look at targets for deep-space rideshare missions with smallsats). I've used it with the poliastro interface to the NASA API for NEOs. Publishing code that I wrote at work can take a while, but I can make it happen if there is interest.

**Comment by Juanlu001** [comment]
Hi @piscisvolans, thanks a lot for chiming in! Of course there is interest :grin: It doesn't even have to start as a pull request to poliastro, but it would be great to see it. Not sure about what timing you are talking about, but as the Zen of Python says, `late is better than never`!

**Comment by Juanlu001** [comment]
Hah, `*now is better than never`, I was too enthusiastic :)

**Comment by IngenieroDeAviones** [comment]
Happy Christmas, everyone!!! :santa: :christmas_tree: :gift: 

I am trying to write a pork-chop that plots the necessary deltaV. For the moment I am just computing de solution to Lambert's problem for each possible time combination, however got some python exceptions.

If anyone has an idea on how to deal with these exceptions or build the algorithm I would appreciate.

Best regards :smile: 

```python

"""
All necessary modules are imported
"""

from astropy.time import Time
from astropy import units as u

import numpy as np

from poliastro.twobody.orbit import Orbit
from poliastro.bodies import Earth, Mars, Sun
from poliastro.util import time_range
from poliastro.iod.vallado import lambert

"""
We select the time spans for launching and arriving, so
we can iterate then over them. The time format is the
following:

'YYYY-MM-DD HH:MM:SS'
""" 

# LAUNCHING TIME SPAN
t0_launch = Time('2017-09-12 00:00:00', scale='tdb')
tf_launch = Time('2018-10-17 00:00:00', scale='tdb')
tspan_launch = time_range(t0_launch, end=tf_launch)


# ARRIVAL TIME SPAN
t0_arrival = Time('2018-08-28 00:00:00', scale='tdb')
tf_arrival = Time('2020-01-10 00:00:00', scale='tdb')
tspan_arrival = time_range(t0_arrival, end=tf_launch)

"""
We iterate for each possible day of launch a set of
arrival times to solve Lambert's problem.
"""
counter = 1
for t0 in tspan_launch: #For a given launch time
    for tf in tspan_arrival: #For a given arrival time at that launch time
        
        tof = tf - t0 # Compute delta T
        
        earth = Orbit.from_body_ephem(Earth, t0) #Earth position at launch
        mars = Orbit.from_body_ephem(Mars, tf) #Mars position at arrival
        
        r0 = earth.r
        r = mars.r
        
        # For the moment only checking that lambert returns a solution.
        # However got the following exceptions:
        #
        # RuntimeError (maximum number of iterations is reached)
        # ZeroDivisionError (division by zero)
        try:
            (v0, vf), = lambert(Sun.k, r0, r, tof)
        
        except RuntimeError:
            print("Got Runtime at computation {}".format(counter))
        
        except ZeroDivisionError:
            print("Got ZeroDivision at computation {}".format(counter))
            
        counter += 1


```

**Comment by Juanlu001** [comment]
Merry Christmas @IngenieroDeAviones! Those Python exceptions are from cases where the algorithm didn't converge, but there might be some genuine failures in our implementation. The best thing is just pass them, for now. There _might_ be smarter sampling strategies but I would start from the simple one as you are doing.

**Comment by piscisvolans** [comment]
@IngenieroDeAviones Happy New Year! I've had success getting fewer exceptions using the Izzo Lambert solver rather than the Vallado solver. I'm not sure if that's because the Izzo solver converges better or because it just reports bogus solutions rather than raising an exception... but you may want to give it a try:

`from poliastro.iod.izzo import lambert`

**Comment by Juanlu001** [comment]
@piscisvolans That's interesting! If you have the time, we would appreciate if you can open an issue with a case where Vallado fails and Izzo doesn't :grin:

**Comment by Juanlu001** [comment]
@IngenieroDeAviones implemented this in #547 - here there are some screenshots :wink: Hope they end up in the documentation soon.

https://twitter.com/poliastro_py/status/1088546275711246337

You are more than welcome to give it a try!
